Sources/ProjectDescription/Dependencies/Dependency+CarthageRequirement.swift

```diff
@@ -0,0 +1,69 @@
+import Foundation
+
+public extension Dependency {
+    enum CarthageRequirement: Codable, Equatable {
```

Comment:
Nitpick: Would you mind moving this to `Dependency.swift` and remove the extension? Not having to jump between files to glance through the models related to `Dependency` results in a better developer experience.

Reply:
You suggested me opposite thing [here](https://github.com/tuist/tuist/pull/2060#discussion_r539189717). Could you explain me a difference?

Reply:
In that case, we only have the extension option because the models are defined in `TuistCore`, and the loading method that we are adding lives in `TuistLoader`. In this case, everything is part of the same framework, so using an extension seems a bit unnecessary and the compiler will be happier about it because it'll be able to parse things faster.

Reply:
ok I have here a few options: 

1.

```swift
public enum Dependency: Codable, Equatable {
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)

    // MARK: - CarthageRequirement

    enum CarthageRequirement: Codable, Equatable {
        case exact(Version)
        case upToNextMajor(Version)
        case upToNextMinor(Version)
        case branch(String)
        case revision(String)

        private enum Kind: String, Codable {
            case exact
            case upToNextMajor
            case upToNextMinor
            case branch
            case revision
        }

        private enum CodingKeys: String, CodingKey {
            case kind
            case version
            case branch
            case revision
        }

        public init(from decoder: Decoder) throws {
            //
        }

        public func encode(to encoder: Encoder) throws {
            //
        }
    }
}
```

2.

```swift
public enum Dependency: Codable, Equatable {
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)
}

// MARK: - CarthageRequirement

extension Dependency {
    enum CarthageRequirement: Codable, Equatable {
        case exact(Version)
        case upToNextMajor(Version)
        case upToNextMinor(Version)
        case branch(String)
        case revision(String)
        
        private enum Kind: String, Codable {
            case exact
            case upToNextMajor
            case upToNextMinor
            case branch
            case revision
        }

        private enum CodingKeys: String, CodingKey {
            case kind
            case version
            case branch
            case revision
        }

        public init(from decoder: Decoder) throws {
            //
        }
        
        public func encode(to encoder: Encoder) throws {
            //
        }
    }
}
```

3. Similar to `ProjectDescription.Setup`.

```swift
public enum Dependency: Codable, Equatable {
    enum CarthageRequirement: Codable, Equatable {
        case exact(Version)
        case upToNextMajor(Version)
        case upToNextMinor(Version)
        case branch(String)
        case revision(String)
    }
    
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)
}

extension Dependency.CarthageRequirement {
    private enum Kind: String, Codable {
        case exact
        case upToNextMajor
        case upToNextMinor
        case branch
        case revision
    }
    
    private enum CodingKeys: String, CodingKey {
        case kind
        case version
        case branch
        case revision
    }
    public init(from decoder: Decoder) throws {
        //
    }
    
    public func encode(to encoder: Encoder) throws {
        //
    }
}
```

4. I have started to think about moving `CarthageRequirement` outside of `Dependency`. 

4a - `CarthageRequirement` and `Dependency` same file
4b- `CarthageRequirement` and `Dependency` separate files

```swift
public enum Dependency: Codable, Equatable {
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)
}

public enum CarthageRequirement: Codable, Equatable {
    case exact(Version)
    case upToNextMajor(Version)
    case upToNextMinor(Version)
    case branch(String)
    case revision(String)
    
    private enum Kind: String, Codable {
        case exact
        case upToNextMajor
        case upToNextMinor
        case branch
        case revision
    }
    
    private enum CodingKeys: String, CodingKey {
        case kind
        case version
        case branch
        case revision
    }
    public init(from decoder: Decoder) throws {
        //
    }
    
    public func encode(to encoder: Encoder) throws {
        //
    }
}
```